### ISSUE TITLE
Consider automatically syncing Blaze and CMake builds

### ISSUE THREAD

**Issue by burmako** [openxla/stablehlo#1581]
Currently, we have to maintain both builds manually, but it would be nice if that could be automated. E.g. IREE folks have a script that does this.

**Comment by GleasonK** [comment]
Is that script public? Can we link it here?

**Comment by burmako** [comment]
@kuhar Can you provide a brief overview of how this works in IREE?

**Comment by kuhar** [comment]
Iree defines bazel and cmake functions to define targets. These generally have matching signatures so that we can generate cmake based off bazel, using a script: https://github.com/openxla/iree/tree/main/build_tools/bazel_to_cmake.

There are also some escape hooks that make this work we need some special cmake (e.g., preserve all cmake below/above line) or when we reference third-party targets whose target don't map 1:1 (e.g., https://github.com/openxla/iree/blob/571f28a87ba01482c8b24b87e4d2a1909ca61f1c/build_tools/bazel_to_cmake/bazel_to_cmake_targets.py#L102-L113).

I believe that @GMNGeoffrey should know the technical details best.

**Comment by kuhar** [comment]
Some concrete example of matching targets below. Note that the cmake is derived from the bazel file, modulo the prefix:
-  https://github.com/openxla/iree/blob/571f28a87ba01482c8b24b87e4d2a1909ca61f1c/compiler/src/iree/compiler/InputConversion/StableHLO/Preprocessing/BUILD.bazel
-  https://github.com/openxla/iree/blob/571f28a87ba01482c8b24b87e4d2a1909ca61f1c/compiler/src/iree/compiler/InputConversion/StableHLO/Preprocessing/CMakeLists.txt

**Comment by GMNGeoffrey** [comment]
Yeah we've had pretty good success with this style of transformation. This makes the automated mapping almost entirely a syntactic thing. IMO it also makes for much more readable CMake files since you make them semi-declarative.

When we looked into this @ScottTodd found https://github.com/google/bazel-to-cmake, which attempts the much more ambitious task of fully mapping Bazel semantics to CMake. We found that it didn't actually work at all (and you can see that it only ever got one commit). Manually mapping Starlark rules to CMake functions makes the technically complicated thing manual and automates the tedious everyday dependency updates. IMO, barring a dedicated staffed team, this is the much more fruitful approach for this sort of thing.

**Comment by ScottTodd** [comment]
Here's some more history on IREE's bazel_to_cmake conversion tool:

* https://groups.google.com/g/iree-discuss/c/Y_Z2n5A-ofY (background, similar pain points, brainstorming, implementation progress)
* https://github.com/openxla/iree/pull/556 (initial PR adding the converter tool)

As Geoffrey mentions, the heavy lifting is done by some declarative CMake functions that behave like Bazel rules (e.g. `cc_library`, `cc_binary`). IREE's versions of these are [`iree_cc_library.cmake`](https://github.com/openxla/iree/blob/main/build_tools/cmake/iree_cc_library.cmake), [`iree_cc_binary.cmake`](https://github.com/openxla/iree/blob/main/build_tools/cmake/iree_cc_binary.cmake), etc.

We have documentation at the top of https://github.com/openxla/iree/blob/main/build_tools/bazel_to_cmake/bazel_to_cmake.py

The critical implementation detail matches the other bazel-to-cmake project: https://github.com/openxla/iree/blob/5c38bcc7068f486733584bd251c9596059c31b7f/build_tools/bazel_to_cmake/bazel_to_cmake_converter.py#L890

* Each `BUILD.bazel` file is ran through `exec` (remember, Bazel uses starlark, which is _close enough_ to Python), with a dictionary of functions to call for each top level function. For example, when a `native_test` function is encountered in a `BUILD.bazel` file, [this block of code](https://github.com/openxla/iree/blob/5c38bcc7068f486733584bd251c9596059c31b7f/build_tools/bazel_to_cmake/bazel_to_cmake_converter.py#L775-L798) runs, parsing the parameters of the function and appending some text to the converted file.

**Comment by GMNGeoffrey** [comment]
I think we should consider moving the core bazel-to-cmake converter out of IREE and into another repo. We do want to avoid a big maintenance burden though. Stella already did some cleanup to make it generalize a bit and is using it in another OpenXLA project that depends on IREE. Obviously StableHLO can't take a dep on IREE though.
